pkg/agent/cniserver/ipam/ipam_delegator.go

```diff
@@ -28,8 +28,9 @@ import (
 )
```

Comment:
But secondary network uses "pkg/agent/secondarynetwork/ipam". Why we need to register Whereabouts in cniserver/ipam?

Reply:
It uses both the IPAM packages @jianjuns.
"pkg/agent/secondarynetwork/ipam"(whereabouts) is used to get the IPAddress from a specific subnet based on the virtual network provided on the Pod spec Annotation. But the IPaddress assignment to the VF interface is actually handled using cniserver/ipam APIs (configureInterface). 

I'm looking to confirm that this fix actually address the issue. But have some lab connectivity issues since yesterday evening and I'm working with my peers at Santa clara lab to recover it. 

Once the lab connectivity is restored today, I will retest this and confirm.

Reply:
handleAddUpdatePod()
    |   Secondary Network Pod Controller.
     configureSecondaryNetwork(pod, networklist, podCNIInfo) 
        |   Get the virtual network specific network-attachment-definition and the CNI config info (as below)
        |   read from network-attachment-def - Config:  { "cniVersion": "0.3.0", "type": "antrea", "ipam": { "type": "whereabouts", "datastore": "kubernetes", "kubernetes": { "kubeconfig": "/host/etc/cni/net.d/whereabouts.d/whereabouts.kubeconfig" }, "range": "146.5.16.0/24" } }
        |   Call configureSecondaryInterface() with Pod, network and CNI information.
        configureSecondaryInterface (pod, netinfo, podCNIInfo, cniconfig)
             |  Build whereabouts cmdArgs and exec Secondary network’s ipam.GetIPAMSubnetAddress(cniconfig, cmdArgs) which uses whereabouts to get an IPaddress allocated from the specified subnet. (Note: this is Secondary network IPAM API).
             |  Build IPAM result and call configureSriovAsSecondaryInterface(pod, netinfo, podCNIInfo, result) to configure SRIOV interface into pod scope for Secondary network need (IP retrieved using whereabouts is passed to this function call).
             configureSriovAsSecondaryInterface(pod, netinfo, podCNIInfo, result)
                  |  cniServer.GetPodConfigurator().ConfigureSriovSecondaryInterface(containerInfo.PodName, containerInfo.PodNameSpace, containerInfo.ContainerID, containerInfo.ContainerNetNS, netinfo.InterfaceName, containerInfo.MTU, podSriovVFDeviceID[0], result);
                   ConfigureSriovSecondaryInterface()
                        |  pc.ifConfigurator.configureContainerLink(podName, podNameSpace, containerID, containerNetNS, containerIFDev, mtu, "", podSriovVFDeviceID, result, nil)
                       |  pc.ifConfigurator.advertiseContainerAddr(containerNetNS, containerIface.Name, result)
                         configureContainerLink()
                             |  ic.configureContainerSriovLink()
                             configureContainerSriovLink()
                                  |  Talk to Intel SRIOV device plugin and get the VF resource information from the pool (specific to the Pod).
                                  |  Call moveVFtoContainerNS(vfIFName, containerID, containerNetNS, containerIfaceName, mtu, result)
                                  moveVFtoContainerNS()
                                        |  Call CNIServer’s ipam.ConfigureIface(containerIface.Name, result) to configure the “given” IP (retrieved from whereabouts CNI) in Pod context)
                                        ipam.ConfigureIface(containerIface.Name, result)